mkosi/gentoo.py

```diff
@@ -0,0 +1,510 @@
+        else:
+            self.arch, kimg_path = self.ARCHITECTURES["x86_64"]
+
+        # GENTOO_UPSTREAM : we only support systemd profiles! and only the no-multilib flaivour , for now;
+        # GENTOO_UPSTREAM : wait for fix upstream: https://bugs.gentoo.org/792081
+        # GENTOO_TODO     : add args.profile switch for this? (multilib vs. nomultilib)
+        self.arch_profile = os.path.join("profiles/default/linux", self.arch, args.release, "no-multilib/systemd")
+
+        self.pkgs_sys = {"@world"}
+
+        self.pkgs_fs = {"sys-fs/dosfstools"}
+        if args.output_format in (OutputFormat.subvolume, OutputFormat.gpt_btrfs):
+            self.pkgs_fs.add("sys-fs/btrfs-progs")
+        elif args.output_format == OutputFormat.gpt_xfs:
+            self.pkgs_fs.add("sys-fs/xfsprogs")
+        elif args.output_format == OutputFormat.gpt_squashfs:
+            self.pkgs_fs.add("sys-fs/squashfs-tools")
+
+        if args.encrypt:
+            self.pkgs_fs.add("cryptsetup")
+            self.pkgs_fs.add("device-mapper")
+
+        self.grub_platforms = set()
+        if not do_run_build_script and args.bootable:
+            if args.esp_partno:
+                self.pkgs_boot = [{"sys-kernel/installkernel-systemd-boot"}]
+            elif args.bios_partno:
+                self.pkgs_boot = [{"sys-boot/grub"}]
+                self.grub_platforms = {"coreboot", "qemu", "pc"}
+                os.environ["GRUB_PLATFORMS"] = " ".join(self.grub_platforms)
+
+            self.pkgs_boot.append(
+                {
+                    "sys-kernel/gentoo-kernel-bin",
+                    "sys-firmware/edk2-ovmf",
+                    # "sys-kernel/linux-firmware",
+                }
+            )
+        self.fetch_fix_stage3(root)
+        self.set_profile(args, root)
+        self.set_default_repo(root)
+        self.unmask_arch(root)
+        self.whitelist_licenses(root)
+        self.provide_patches(args, root)
+        self.set_useflags(root)
+        self.gentoo_mkosi_conf(root)
+        self.invoke_emerge(args, root, inside_stage3=False, actions={"--sync"})
+        self.baselayout(args, root)
+        self.update_stage3(args, root)
+
+        if "build-script" in ARG_DEBUG:
+            self.invoke_emerge(args, root, actions={"--info"})
+
+    def fetch_fix_stage3(self, root: str) -> None:
+        """usrmerge tracker bug: https://bugs.gentoo.org/690294"""
+
+        # e.g.: http://distfiles.gentoo.org/releases/amd64/autobuilds/latest-stage3.txt
+        stage3tsf_path_url = urllib.parse.urljoin(
+            self.portage_cfg["GENTOO_MIRRORS"].partition(" ")[0],
+            f"releases/{self.arch}/autobuilds/latest-stage3.txt",
+        )
+        stage3_tar = ""
+        # GENTOO_UPSTREAM: wait for fix upstream: https://bugs.gentoo.org/792081
+        # and more... so we can gladly escape all this hideousness!
+        with urllib.request.urlopen(stage3tsf_path_url) as r:
```

Comment:
Doesn't .urlopen() have some option to spit out decoded text?

Reply:
it does have `read()` and `readlines()` i will use those.
i think i copy-past'ed this from `archlinux` in `mkosi/__init__.py`...